res/controllers/Pioneer-DDJ-400-script.js

```diff
@@ -0,0 +1,741 @@
+    var idx = 0;
+
+    for (var i = 0; i < this.tempoRanges.length; i++) {
+        if (currRange === this.tempoRanges[i]) {
+            idx = (i + 1) % this.tempoRanges.length;
+            break;
+        }
+    }
+    engine.setValue(group, "rateRange", this.tempoRanges[idx]);
+};
+
+//
+// Jog wheels
+//
+
+PioneerDDJ400.jogTurn = function(channel, _control, value, _status, group) {
+    var deckNum = channel + 1;
+    // wheel center at 64; <64 rew >64 fwd
+    var newVal = value - 64;
+
+    // loop_in / out adjust
+    var loopEnabled = engine.getValue(group, "loop_enabled");
+    if (loopEnabled > 0) {
+        if (PioneerDDJ400.loopAdjustIn[channel]) {
+            newVal = newVal * PioneerDDJ400.loopAdjustMultiply + engine.getValue(group, "loop_start_position");
+            engine.setValue(group, "loop_start_position", newVal);
+            return;
+        }
+        if (PioneerDDJ400.loopAdjustOut[channel]) {
+            newVal = newVal * PioneerDDJ400.loopAdjustMultiply + engine.getValue(group, "loop_end_position");
+            engine.setValue(group, "loop_end_position", newVal);
+            return;
+        }
+    }
+
+    if (engine.isScratching(deckNum)) {
+        engine.scratchTick(deckNum, newVal);
+    } else { // fallback
+        engine.setValue(group, "jog", newVal * this.bendScale);
+    }
+};
+
+
+PioneerDDJ400.jogSearch = function(_channel, _control, value, _status, group) {
+    // "highspeed" (scaleup value) pitch bend
+    var newVal = (value - 64) * this.highspeedScale;
+    engine.setValue(group, "jog", newVal);
+};
+
+PioneerDDJ400.jogTouch = function(channel, _control, value) {
+    var deckNum = channel + 1;
+
+    // skip scratchmode if we adjust the loop points
+    if (PioneerDDJ400.loopAdjustIn[channel] || PioneerDDJ400.loopAdjustOut[channel]) {
+        return;
+    }
+
+    // on touch jog with vinylmode enabled -> enable scratchmode
+    if (value !== 0 && this.vinylMode) {
+        engine.scratchEnable(deckNum, 720, 33+1/3, this.alpha, this.beta);
+    } else {
+        // on release jog (value === 0) disable pitch bend mode or scratch mode
+        engine.scratchDisable(deckNum);
+    }
+};
+
+//
+// SHIFT
+//
+
+PioneerDDJ400.shiftPressed = function(channel, _control, value) {
+    this.shiftState[channel] = value;
+    this.performancePads.toggleShiftLights(channel, value);
+};
+
+//
+// Pad mode buttons
+//
+
+PioneerDDJ400.setPadmode = function(channel, control, value) {
+    if (value === 0x7F) {
+        PioneerDDJ400.performancePads.state[channel] = control;
+    }
+};
+
+//
+// Beat Jump mode
+//
+
+PioneerDDJ400.beatjumpPadPressed = function(_channel, control, value, _status, group) {
+    if (value === 0) {
+        return;
+    }
+    engine.setValue(group, "beatjump_size", Math.abs(PioneerDDJ400.beatjumpPad[control]));
+    engine.setValue(group, "beatjump", PioneerDDJ400.beatjumpPad[control]);
+};
+
+PioneerDDJ400.beatjumpShiftUp = function(_channel, control, value, _status, group) {
```

Comment:
```suggestion
PioneerDDJ400.increaseBeatjumpSizes = function(_channel, control, value, _status, group) {
```
It took me a while to understand what this function does.